File: web-sdk/src/data/services/yalo-media/yalo-media-service-remote.test.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { YaloMediaServiceRemote } from './yalo-media-service-remote';
import { Ok, Err } from '@domain/common/result';
import type { TokenRepository } from '@data/repositories/token/token-repository';

const makeTokenRepository = (token = 'test-token'): TokenRepository => ({
  getToken: vi.fn().mockResolvedValue(new Ok(token)),
});

const makeUploadResponse = () => ({
  id: 'yalo_63cfd924-19dc-455a-b755-c949e62fc4e7',
  signed_url: 'https://storage.googleapis.com/example/file.jpeg',
  original_name: 'photo.jpeg',
  type: 'image',
  metadata: { user_id: 'abc123' },
  created_at: '2026-03-19T15:56:21.629107809Z',
  expires_at: '2026-03-19T16:56:21.190587072Z',
});

describe('YaloMediaServiceRemote', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe('uploadMedia', () => {
    it('returns Ok with MediaUploadResponse on 201', async () => {
      const responseBody = makeUploadResponse();
      vi.stubGlobal(
        'fetch',
        vi.fn().mockResolvedValue({
          status: 201,
          json: vi.fn().mockResolvedValue(responseBody),
        })
      );

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        makeTokenRepository()
      );
      const file = new File(['fake image'], 'photo.jpeg', {
        type: 'image/jpeg',
      });
      const result = await service.uploadMedia(file);

      expect(result.ok).toBe(true);
      if (result.ok) {
        expect(result.value.id).toBe(
          'yalo_63cfd924-19dc-455a-b755-c949e62fc4e7'
        );
        expect(result.value.signed_url).toBe(
          'https://storage.googleapis.com/example/file.jpeg'
        );
        expect(result.value.original_name).toBe('photo.jpeg');
        expect(result.value.type).toBe('image');
      }
    });

    it('POSTs to correct URL with Authorization header and FormData', async () => {
      const fetchSpy = vi.fn().mockResolvedValue({
        status: 201,
        json: vi.fn().mockResolvedValue(makeUploadResponse()),
      });
      vi.stubGlobal('fetch', fetchSpy);

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        makeTokenRepository('my-token')
      );
      const file = new File(['data'], 'photo.jpeg', { type: 'image/jpeg' });
      await service.uploadMedia(file);

      expect(fetchSpy).toHaveBeenCalledWith(
        'https://api.example.com/all/media',
        expect.objectContaining({
          method: 'POST',
          headers: { Authorization: 'Bearer my-token' },
        })
      );

      const [, init] = fetchSpy.mock.calls[0];
      expect(init.body).toBeInstanceOf(FormData);
      expect(init.body.get('file')).toBeInstanceOf(File);
    });

    it('returns Err on non-201 status', async () => {
      vi.stubGlobal(
        'fetch',
        vi.fn().mockResolvedValue({ status: 400 })
      );

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        makeTokenRepository()
      );
      const file = new File(['data'], 'photo.jpeg', { type: 'image/jpeg' });
      const result = await service.uploadMedia(file);

      expect(result.ok).toBe(false);
      if (!result.ok)
        expect(result.error.message).toBe('Failed to upload media: 400');
    });

    it('returns Err when token fetch fails', async () => {
      const tokenRepo: TokenRepository = {
        getToken: vi
          .fn()
          .mockResolvedValue(new Err(new Error('auth failed'))),
      };

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        tokenRepo
      );
      const file = new File(['data'], 'photo.jpeg', { type: 'image/jpeg' });
      const result = await service.uploadMedia(file);

      expect(result.ok).toBe(false);
      if (!result.ok) expect(result.error.message).toBe('auth failed');
    });

    it('returns Err when fetch throws', async () => {
      vi.stubGlobal(
        'fetch',
        vi.fn().mockRejectedValue(new Error('Network error'))
      );

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        makeTokenRepository()
      );
      const file = new File(['data'], 'photo.jpeg', { type: 'image/jpeg' });
      const result = await service.uploadMedia(file);

      expect(result.ok).toBe(false);
      if (!result.ok) expect(result.error.message).toBe('Network error');
    });

    it('wraps non-Error thrown values', async () => {
      vi.stubGlobal('fetch', vi.fn().mockRejectedValue('string error'));

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        makeTokenRepository()
      );
      const file = new File(['data'], 'photo.jpeg', { type: 'image/jpeg' });
      const result = await service.uploadMedia(file);

      expect(result.ok).toBe(false);
      if (!result.ok) expect(result.error.message).toBe('string error');
    });
  });

  describe('downloadMedia', () => {
    it('returns Ok with Blob on 200', async () => {
      const blob = new Blob(['image data'], { type: 'image/jpeg' });
      vi.stubGlobal(
        'fetch',
        vi.fn().mockResolvedValue({
          ok: true,
          blob: vi.fn().mockResolvedValue(blob),
        })
      );

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        makeTokenRepository()
      );
      const result = await service.downloadMedia(
        'https://example.com/image.jpg'
      );

      expect(result.ok).toBe(true);
      if (result.ok) expect(result.value).toBeInstanceOf(Blob);
    });

    it('returns Err on non-200 response', async () => {
      vi.stubGlobal(
        'fetch',
        vi.fn().mockResolvedValue({ ok: false, status: 404 })
      );

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        makeTokenRepository()
      );
      const result = await service.downloadMedia(
        'https://example.com/image.jpg'
      );

      expect(result.ok).toBe(false);
      if (!result.ok)
        expect(result.error.message).toBe('Failed to download media: 404');
    });

    it('returns Err when fetch throws', async () => {
      vi.stubGlobal(
        'fetch',
        vi.fn().mockRejectedValue(new Error('Network error'))
      );

      const service = new YaloMediaServiceRemote(
        'https://api.example.com',
        makeTokenRepository()
      );
      const result = await service.downloadMedia(
        'https://example.com/image.jpg'
      );

      expect(result.ok).toBe(false);
      if (!result.ok) expect(result.error.message).toBe('Network error');
    });
  });
});

File: flutter-sdk/test/data/services/yalo_media/yalo_media_service_remote_test.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'dart:convert';
import 'dart:typed_data';

import 'package:yalo_chat_flutter_sdk/src/common/result.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_media/media_upload_response.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_media/yalo_media_service_remote.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_message_auth/token_entry.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_message_auth/yalo_message_auth_service.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart';
import 'package:image_picker/image_picker.dart';
import 'package:mocktail/mocktail.dart';

class MockYaloMessageAuthService extends Mock
    implements YaloMessageAuthService {}

class MockClient extends Mock implements Client {}

TokenEntry _makeTokenEntry(String token) => TokenEntry(
  accessToken: token,
  userId: 'user-123',
  refreshToken: '',
  expiresAt: DateTime.now().add(const Duration(hours: 1)),
);

const _responseBody = '''
{
  "id": "yalo_63cfd924-19dc-455a-b755-c949e62fc4e7",
  "signed_url": "https://storage.googleapis.com/example/file.jpeg",
  "original_name": "photo.jpeg",
  "type": "image",
  "metadata": {"user_id": "abc123"},
  "created_at": "2026-03-19T15:56:21.629107809Z",
  "expires_at": "2026-03-19T16:56:21.190587072Z"
}
''';

void main() {
  setUpAll(() {
    registerFallbackValue(Request('GET', Uri.parse('')));
  });

  group('YaloMediaServiceRemote', () {
    late MockYaloMessageAuthService mockAuthService;
    late MockClient mockClient;
    late YaloMediaServiceRemote service;

    const baseUrl = 'https://api.example.com';
    const token = 'test-token';

    setUp(() {
      mockAuthService = MockYaloMessageAuthService();
      mockClient = MockClient();
      service = YaloMediaServiceRemote(
        baseUrl: baseUrl,
        authService: mockAuthService,
        httpClient: mockClient,
      );
    });

    XFile makeXFile() => XFile.fromData(
      utf8.encode('fake image bytes'),
      name: 'photo.jpeg',
      mimeType: 'image/jpeg',
    );

    StreamedResponse makeStreamedResponse(int statusCode, String body) =>
        StreamedResponse(Stream.value(utf8.encode(body)), statusCode);

    group('uploadMedia', () {
      test('propagates auth error without calling send', () async {
        when(
          () => mockAuthService.auth(),
        ).thenAnswer((_) async => Result.error(Exception('auth failed')));

        final result = await service.uploadMedia(makeXFile());

        expect(result, isA<Error<MediaUploadResponse>>());
        verifyNever(() => mockClient.send(any()));
      });

      test('POSTs to correct URL with Authorization header', () async {
        when(
          () => mockAuthService.auth(),
        ).thenAnswer((_) async => Result.ok(_makeTokenEntry(token)));
        when(
          () => mockClient.send(any()),
        ).thenAnswer((_) async => makeStreamedResponse(201, _responseBody));

        await service.uploadMedia(makeXFile());

        final captured = verify(() => mockClient.send(captureAny())).captured;
        final request = captured.first as MultipartRequest;

        expect(request.url.toString(), equals('$baseUrl/all/media'));
        expect(request.headers['Authorization'], equals('Bearer $token'));
        expect(request.files, hasLength(1));
        expect(request.files.first.field, equals('file'));
      });

      test('returns Ok<MediaUploadResponse> on 201 with valid JSON', () async {
        when(
          () => mockAuthService.auth(),
        ).thenAnswer((_) async => Result.ok(_makeTokenEntry(token)));
        when(
          () => mockClient.send(any()),
        ).thenAnswer((_) async => makeStreamedResponse(201, _responseBody));

        final result = await service.uploadMedia(makeXFile());

        expect(result, isA<Ok<MediaUploadResponse>>());
        final response = (result as Ok<MediaUploadResponse>).result;
        expect(
          response.id,
          equals('yalo_63cfd924-19dc-455a-b755-c949e62fc4e7'),
        );
        expect(
          response.signedUrl,
          equals('https://storage.googleapis.com/example/file.jpeg'),
        );
        expect(response.originalName, equals('photo.jpeg'));
        expect(response.type, equals('image'));
        expect(response.metadata['user_id'], equals('abc123'));
      });

      test('returns Error on non-201 response', () async {
        when(
          () => mockAuthService.auth(),
        ).thenAnswer((_) async => Result.ok(_makeTokenEntry(token)));
        when(
          () => mockClient.send(any()),
        ).thenAnswer((_) async => makeStreamedResponse(400, ''));

        final result = await service.uploadMedia(makeXFile());

        expect(result, isA<Error<MediaUploadResponse>>());
      });

      test('returns Error when client throws', () async {
        when(
          () => mockAuthService.auth(),
        ).thenAnswer((_) async => Result.ok(_makeTokenEntry(token)));
        when(
          () => mockClient.send(any()),
        ).thenThrow(Exception('network error'));

        final result = await service.uploadMedia(makeXFile());

        expect(result, isA<Error<MediaUploadResponse>>());
      });
    });

    group('downloadMedia', () {
      const imageUrl = 'https://example.com/image.jpg';
      final imageBytes = Uint8List.fromList([1, 2, 3, 4]);

      test('returns Ok<Uint8List> on 200', () async {
        when(
          () => mockClient.get(Uri.parse(imageUrl)),
        ).thenAnswer((_) async => Response.bytes(imageBytes, 200));

        final result = await service.downloadMedia(imageUrl);

        expect(result, isA<Ok<Uint8List>>());
        expect((result as Ok<Uint8List>).result, equals(imageBytes));
      });

      test('returns Error on non-200 response', () async {
        when(
          () => mockClient.get(Uri.parse(imageUrl)),
        ).thenAnswer((_) async => Response('', 404));

        final result = await service.downloadMedia(imageUrl);

        expect(result, isA<Error<Uint8List>>());
      });

      test('returns Error when client throws', () async {
        when(
          () => mockClient.get(Uri.parse(imageUrl)),
        ).thenThrow(Exception('network error'));

        final result = await service.downloadMedia(imageUrl);

        expect(result, isA<Error<Uint8List>>());
      });
    });
  });
}
